### ISSUE TITLE
Support for the Rack Websocket extension draft?

### ISSUE THREAD

**Issue by boazsegev** [ohler55/agoo#6]
Hi,

I just discovered this project and it looks amazing!

I love the routing layer (I wanted to add something similar to iodine but haven't gotten around to it just yet, I want to make it all in C and support the `/path/(:param)` routing convention).

I was wondering if there's any chance I could interest you in supporting the [Rack Websocket extension draft](https://github.com/boazsegev/iodine/blob/master/SPEC-Websocket-Draft.md).

Specifically I would love it if your server supported the Websocket upgrade (using `req['upgrade.websocket?']`) and the *optional* pub/sub extensions (`subscribe` / `publish` / `unsubscribe`)?

There was [a discussion about the draft](https://github.com/boazsegev/iodine/issues/6) between different Ruby server maintainers a while back (it [started here](https://github.com/rack/rack/issues/1093) and moved [here](https://github.com/boazsegev/iodine/issues/6)), but nobody implemented the draft at the end of it (except me).

I believe these things should be separated from the application layer, as they seem to be network related concerns that should be embedded in the server.

The Websocket upgrade is relatively simple to implement (you can use C code from [iodine](https://github.com/boazsegev/iodine) or from [facil.io](http://facil.io), such as the [Websocket parser / formatter](https://github.com/boazsegev/facil.io/blob/6cf44b212723693a976d4c9f4519f638cdd8ef30/lib/facil/http/parsers/websocket_parser.h)).

The optional pub/sub can take a lot of work if you're supporting a cluster mode, but you don't have to implement it (it's an extension to the draft, not a requirement)... however, it would be beneficial to separate pub/sub logic from the application layer, especially if Redis or another network pub/sub service is attached.

Kindly,
   Bo.

**Comment by ohler55** [comment]
Ahh, The Iodine owner. You've got a fast server there. I like the multiple worker approach for when a shared state is not needed. Agoo and Iodine are neck and neck on performance.

I do plan on adding support for websockets. I have a library I wrote and have used on another projec (piperpushcache.com)t. I wasn't sure how much demand there was for it so I wasn't in a big hurry. I'll look at the links you provided and learn more about where the demand is. I had read somewhere that the plan to support websockets was to hijack the connection and then leave it up to the app. I'm with you in thinking that is a cop out and some common approach in the server would be better. 

My favorite pub/sub is NATS. A friend of mine wrote that and I have used it on projects at work and for play. It has a Ruby API as well C and quite a few others. I'm not sure how that relates to websockets with support or at least not sure how it would directly relate. I could certainly see another gem or bit of code that bridges the network pub/sub component to the websockets component. I do like a pub/sub model for websockets though.

I am currently working on two dev stream, Agoo and OpO which is closed source for now. OpO is a fast JSON database. My plan for clustering is to span multiple machines and use the OpO database for shared state. Thats a little ways off though.

I'm glad you reached out. We can take more discussions via email but lets leave this issue open as a place holder for websocket support. I'll update it with comments related to that as I move along.
